graph/patterns/alternate-key.md

```diff
@@ -0,0 +1,130 @@
+
+While it is still possible to use the OData filter, such as
+
+`https://graph.microsoft.com/v1.0/users?$filter=mail eq 'bob@contoso.com'`, the returned result is wrapped in an array that needs to be unpacked.
+
+## Solution
+
+---
+
+OData offers resource addressing via an alternate key using the same parentheses-style convention as for the canonical key, with one difference: single-part alternate keys MUST specify the key property name to unambiguously determine the alternate key. (Note: this is a hypothetical sample)
+
+https://graph.microsoft.com/v1.0/users(0) - Retrieves the employee with ID = 0
+https://graph.microsoft.com/v1.0/users(email='bob@contoso.com') Retrieves the employee with the email matching `bob@contoso.com`
+
+## When to Use this Pattern
+
+---
+
+This pattern works and makes sense when the alternate key is good enough to identify a single resource and provides an useful alternative to the client; while it does not work if the resultset has more than one element. In such case, the workload SHOULD return `400`
+
+## Example
+
+---
+
+The same user identified via the alternate key SSN, the canonical (primary) key ID using the non-canonical long form with specified key property name, and the canonical short form without key property name
+
+Declare `mail` and `ssn` as alternate keys on an entity:
+
+```xml
+<EntityType Name="User">
+   <Key>
+     <PropertyRef Name="id" />
+   </Key>
+   <Property Name="id" Type="Edm.Int32" />
+
+   <Property Name="mail" Type="Edm.String" />
+   <Property Name="ssn" Type="Edm.String" />
+   <Annotation Term="Keys.AlternateKeys">
+      <Collection>
+         <Record>
+            <PropertyValue Property="Key">
+               <Collection>
+                  <Record>
+                     <PropertyValue Property="Name" PropertyPath="mail" />
+                  </Record>
+               </Collection>
+            </PropertyValue>
+         </Record>
+         <Record>
+            <PropertyValue Property="Key">
+               <Collection>
+                  <Record>
+                     <PropertyValue Property="Name" PropertyPath="ssn" />
+                  </Record>
+               </Collection>
+            </PropertyValue>
+         </Record>
+      </Collection>
+   </Annotation>
+</EntityType>
+```
+
+1. Get a specific resource through `$filter`:
+
+```http
+GET https://graph.microsoft.com/v1.0/users/?$filter=(ssn eq '123-45-6789')
+```
+
+```json
+{
+  "value": [
+    {
+      "givenName": "Bob",
+      "jobTitle": "Retail Manager",
+      "mail": "bob@contoso.com",
+      "mobilePhone": "+1 425 555 0109",
+      "officeLocation": "18/2111",
+      "preferredLanguage": "en-US",
+      "surname": "Vance",
+      "userPrincipalName": "bob@contoso.com",
+      "id": "1a89ade6-9f59-4fea-a139-23f84e3aef66"
+    }
+  ]
+}
+```
+
+2. Get a specific resource through its primary key, and then through the two alternate keys:
```

Comment:
Get a specific resource either through its primary key, or through the two alternate keys